File: dash_docs/chapters/clientside_callbacks/examples/graph_update_fe_be.R
library(dash)
library(dashHtmlComponents)
library(dashCoreComponents)
library(plotly)
library(jsonlite)

app <- Dash$new()

df = read.csv("datasets/gapminderDataFiveYear.csv")

available_countries = unique(df$country)

app$layout(htmlDiv(list(
  dccGraph(
    id='clientside-graph'
  ),
  dccStore(
    id='clientside-figure-store',
    data = list(
      list(
        'x' = df[df$country == 'Canada', "year"],
        'y' = df[df$country == 'Canada', "pop"]
      )
    )
  ),
  'Indicator',
  dccDropdown(
    id='clientside-graph-indicator',
    options = list(
      list('label' = 'Population', 'value' = 'pop'),
      list('label' = 'Life Expectancy', 'value' = 'lifeExp'),
      list('label' = 'GDP per Capita', 'value' = 'gdpPercap')
    ),
    value='pop'
  ),
  'Country',
  dccDropdown(
    id='clientside-graph-country',
    options=lapply(available_countries, function(x) {
      return(list('label' = x, 'value' = x))
    }),
    value = 'Canada'
  ),
  'Graph Scale',
  dccRadioItems(
    id='clientside-graph-scale',
    options=list(
      list('label' = 'Linear', 'value' = 'linear'),
      list('label' = 'Log', 'value' = 'log')
    ),
    value = 'linear'
  ),
  htmlHr(),
  htmlDetails(
    list(
      htmlSummary('Contents of figure storage'),
      dccMarkdown(
        id='clientside-figure-json'
        )
    )
  )
)))

app$callback(
  output('clientside-figure-store', 'data'),
  params = list(
    input('clientside-graph-indicator', 'value'),
    input('clientside-graph-country', 'value')
  ),
  function(indicator, selected_country) {
    dff = df[df$country == selected_country,]
    return(list(list(
      'x' = dff[, "year"],
      'y' = dff[, indicator],
      'mode' = 'markers'
    )))
  }
)

# Clientside callback that interacts with dccStore
app$callback(
  output('clientside-graph', 'figure'),
  params = list(
    input('clientside-figure-store', 'data'),
    input('clientside-graph-scale', 'value')
  ),
  clientsideFunction(
    namespace = 'clientside_examples',
    function_name = 'update_graph'
  )
)

app$callback(
  output('clientside-figure-json', 'children'),
  params = list(
    input('clientside-figure-store', 'data')
  ),
  function(input_data) {
    return(jsonlite::prettify(dash:::to_JSON(input_data)))
  }
)

app$run_server()

File: dash_docs/chapters/clientside_callbacks/examples/graph_update_fe_be.py
import dash
from dash.dependencies import Input, Output
import dash_core_components as dcc
import dash_html_components as html
import pandas as pd

import json

external_stylesheets = ['https://codepen.io/chriddyp/pen/bWLwgP.css']

app = dash.Dash(__name__, external_stylesheets=external_stylesheets)

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv')

available_countries = df['country'].unique()

app.layout = html.Div([
    dcc.Graph(
        id='clientside-graph'
    ),
    dcc.Store(
        id='clientside-figure-store',
        data=[{
            'x': df[df['country'] == 'Canada']['year'],
            'y': df[df['country'] == 'Canada']['pop']
        }]
    ),
    'Indicator',
    dcc.Dropdown(
        id='clientside-graph-indicator',
        options=[
            {'label': 'Population', 'value': 'pop'},
            {'label': 'Life Expectancy', 'value': 'lifeExp'},
            {'label': 'GDP per Capita', 'value': 'gdpPercap'}
        ],
        value='pop'
    ),
    'Country',
    dcc.Dropdown(
        id='clientside-graph-country',
        options=[
            {'label': country, 'value': country}
            for country in available_countries
        ],
        value='Canada'
    ),
    'Graph scale',
    dcc.RadioItems(
        id='clientside-graph-scale',
        options=[
            {'label': x, 'value': x} for x in ['linear', 'log']
        ],
        value='linear'
    ),
    html.Hr(),
    html.Details([
        html.Summary('Contents of figure storage'),
        dcc.Markdown(
            id='clientside-figure-json'
        )
    ])
])


@app.callback(
    Output('clientside-figure-store', 'data'),
    [Input('clientside-graph-indicator', 'value'),
     Input('clientside-graph-country', 'value')]
)
def update_store_data(indicator, country):
    dff = df[df['country'] == country]
    return [{
        'x': dff['year'],
        'y': dff[indicator],
        'mode': 'markers'
    }]


app.clientside_callback(
    """
    function(data, scale) {
        return {
            'data': data,
            'layout': {
                 'yaxis': {'type': scale}
             }
        }
    }
    """,
    Output('clientside-graph', 'figure'),
    [Input('clientside-figure-store', 'data'),
     Input('clientside-graph-scale', 'value')]
)


@app.callback(
    Output('clientside-figure-json', 'children'),
    [Input('clientside-figure-store', 'data')]
)
def generated_figure_json(data):
    return '```\n'+json.dumps(data, indent=2)+'\n```'


if __name__ == '__main__':
    app.run_server(debug=True)
